File: .cursor/skills/sleuths/sleuth/src/relevance.rs
#[derive(Debug, serde::Deserialize)]
struct RelevanceResponse {
    #[serde(default)]
    relevant_ids: Vec<serde_json::Value>,
}

/// Parse LLM relevance output into zero-based chunk indices.
/// Invalid JSON, wrong shape, or out-of-range ids are ignored; parse failure → empty.
pub fn parse_relevant_ids(response: &str, max_index: usize) -> Vec<usize> {
    let trimmed = strip_code_fences(response.trim());
    let parsed: RelevanceResponse = match serde_json::from_str(trimmed) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("relevance: failed to parse JSON: {e}");
            return vec![];
        }
    };

    let mut ids = Vec::new();
    for v in parsed.relevant_ids {
        let idx = match v {
            serde_json::Value::Number(n) => n.as_u64().and_then(|u| usize::try_from(u).ok()),
            _ => None,
        };
        let Some(idx) = idx else {
            continue;
        };
        if idx > max_index {
            continue;
        }
        if !ids.contains(&idx) {
            ids.push(idx);
        }
    }
    ids.sort_unstable();
    ids
}

fn strip_code_fences(s: &str) -> &str {
    let s = s.trim();
    if let Some(rest) = s.strip_prefix("```json") {
        return rest.strip_suffix("```").unwrap_or(rest).trim();
    }
    if let Some(rest) = s.strip_prefix("```") {
        return rest.strip_suffix("```").unwrap_or(rest).trim();
    }
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_valid_ids() {
        let ids = parse_relevant_ids(r#"{"relevant_ids": [0, 2, 5]}"#, 5);
        assert_eq!(ids, vec![0, 2, 5]);
    }

    #[test]
    fn ignores_out_of_range() {
        let ids = parse_relevant_ids(r#"{"relevant_ids": [0, 99]}"#, 3);
        assert_eq!(ids, vec![0]);
    }

    #[test]
    fn strips_fences() {
        let raw = "```json\n{\"relevant_ids\": [1]}\n```";
        let ids = parse_relevant_ids(raw, 3);
        assert_eq!(ids, vec![1]);
    }

    #[test]
    fn parse_failure_empty() {
        assert!(parse_relevant_ids("not json", 3).is_empty());
    }
}

File: .cursor/skills/sleuths/sleuth/relevance.py
from __future__ import annotations

import re
import sys


def parse_relevant_ids(response: str, max_index: int) -> list[int]:
    """Parse LLM relevance output into zero-based chunk indices."""
    trimmed = _strip_code_fences(response.strip())
    if not trimmed:
        return []

    ids: list[int] = []
    saw_token = False
    for token in trimmed.split(","):
        part = token.strip()
        if not part:
            continue
        saw_token = True
        if not re.fullmatch(r"-?\d+", part):
            print(
                f"relevance: skipping non-numeric token {part!r}",
                file=sys.stderr,
            )
            continue
        idx = int(part)
        if idx < 0 or idx > max_index:
            continue
        if idx not in ids:
            ids.append(idx)

    if saw_token and not ids:
        print("relevance: no valid indices in response", file=sys.stderr)
        return []
    if not saw_token and trimmed:
        print(f"relevance: unparseable response: {trimmed!r}", file=sys.stderr)
        return []

    ids.sort()
    return ids


def _strip_code_fences(s: str) -> str:
    s = s.strip()
    if s.startswith("```json"):
        rest = s[7:]
        if rest.endswith("```"):
            rest = rest[:-3]
        return rest.strip()
    if s.startswith("```"):
        rest = s[3:]
        if rest.endswith("```"):
            rest = rest[:-3]
        return rest.strip()
    return s
